### PR TITLE
batch plugin logs on postgres

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

Addresses #424 and supersedes #431.

This makes log inserts to Postgres from plugins happen in batches every second. Didn't implement batching for other logs but should be easy to do so later if we feel there's a need. Big improvements from this locally. No slow query warnings, and despite the buffer, the logs actually appear faster (since they're not "stuck in traffic").

## Checklist

-   [ ] Updated Settings section in README.md, if settings are affected
-   [ ] Jest tests

### CODE REVIEW

**Review by neilkakkar — APPROVED** [review]
LGTM.

Surprised though that this solves the underlying issue.

In effect, were we making too many queries in succession to postgres? We're still making about 1/second. So does this imply earlier, we were even faster than that?

**Review by mariusandra — CHANGES_REQUESTED** [review]
Very nice! Just two thoughts:
- even though we flush every second, perhaps for posterity it makes sense to also flush this buffer in teardown/shutdown mode? it's yet another stream/buffer/service/etc, so might make sense to control in a similar way...
- it looks like you're always inserting to postgres, even if kafka is enabled.

File: src/worker/vm/extensions/console.ts [hunk 1]
```diff
@@ -17,18 +19,30 @@ function consoleFormat(...args: unknown[]): string {
...
 
 export function createConsole(server: Hub, pluginConfig: PluginConfig): ConsoleExtension {
-    async function consolePersist(type: PluginLogEntryType, ...args: unknown[]): Promise<void> {
+    const consoleBuffer = createBuffer({
+        limit: 100,
+        timeoutSeconds: 1,
+        onFlush: async (logEntries: LogEntryPayload[]) => {
+            await server.db.createPluginLogEntries(pluginConfig, logEntries)
+        },
+    })
```
Comment (Twixes) [inline src/worker/vm/extensions/console.ts:28]: This solution seems very high-level, as it only covers `console` and in a per-plugin fashion – compared to Kafka message queuing, which covers the whole plugin server thread. I think if we do want to go for batching here, it may be cleaner and faster to have `createPluginLogEntry` append to a buffer, which would be flush for the whole the thread. That would avoid the situation where there are also _two_ buffer levels for Kafka-produced `console` logs (initial buffer here, then second buffer in `kafkaProducer.queueMessage`).

**Review by Twixes — CHANGES_REQUESTED** [review]

File: src/utils/db/db.ts [hunk 2]
```diff
@@ -698,52 +706,73 @@ export class DB {
...
+                        topic: KAFKA_PLUGIN_LOG_ENTRIES,
+                        messages: [{ key: parsedEntry.id, value: Buffer.from(JSON.stringify(parsedEntry)) }],
+                    })
+                } catch (e) {
+                    captureException(e)
+                    console.error(parsedEntry)
+                    console.error(e)
+                }
+            }
+            return
```
Comment (yakkomajuri) [inline src/utils/db/db.ts:743]: @mariusandra this should prevent the insert into postgres

File: src/worker/vm/extensions/console.ts (hunk 1 above)
Comment (yakkomajuri) [reply src/worker/vm/extensions/console.ts:28]: It is indeed. I originally came at this for writing a general batching system, but decided to start with this as the bulk of logs comes from plugins and I had to work on other things.

**Comment by yakkomajuri** [comment]
@mariusandra re

```
even though we flush every second, perhaps for posterity it makes sense to also flush this buffer in teardown/shutdown mode? it's yet another stream/buffer/service/etc, so might make sense to control in a similar way...
```

yeah that'd make sense

File: src/worker/vm/extensions/console.ts (hunk 1 above)
Comment (yakkomajuri) [reply src/worker/vm/extensions/console.ts:28]: The system is there to take a batch of entries and insert them so it definitely isn't a major step to write worker-level memory batching. I just took over #431 to get this out quickly

File: src/utils/db/db.ts (hunk 2 above)
Comment (mariusandra) [reply src/utils/db/db.ts:743]: Oh god, that's embarrassing :D. 

In my defence I have a strong cold, but still... :D

**Comment by yakkomajuri** [comment]
huh, these passed locally 🤔 
Screenshot 2021-07-02 at 10 44 43

**Comment by yakkomajuri** [comment]
@Twixes thread-level batching now

**Comment by yakkomajuri** [comment]
friendly ping @Twixes

File: tests/postgres/e2e.test.ts [hunk 3]
```diff
@@ -117,10 +117,10 @@ describe('e2e postgres ingestion', () => {
 
         posthog.capture('custom event', { name: 'hehe', uuid: new UUIDT().toString() })
 
+        // await delayUntilEventIngested(async () =>  )
```
Comment (Twixes) [inline tests/postgres/e2e.test.ts:120]: Why commented out?

File: tests/postgres/vm.test.ts [hunk 4]
```diff
@@ -29,6 +30,7 @@ let closeHub: () => Promise<void>
 beforeEach(async () => {
     ;(Client as any).mockClear()
     ;[hub, closeHub] = await createHub()
+    // hub.db.createPluginLogEntries = jest.fn()
```
Comment (Twixes) [inline tests/postgres/vm.test.ts:33]: Commented out

**Review by Twixes — COMMENTED** [review]
Looks like this works so if unbatched Postgres log insertion is a hair-on-fire problem, it's okay, but still has this thing where Kafka logs are now batched two-fold – first via `addLog` and then via `queueMessage`. Not ideal since that only increases the time in which a log entry gets into ClickHouse, without benefits. Basically ideally `createPluginLogEntries` should queue a Kafka message (which it already does) _or_ queue a Postgres row, instead of `LogsBuffer` creating an additional buffer layer above `createPluginLogEntries`.

**Comment by yakkomajuri** [comment]
That's a good point - `addLog` can just directly throw it in Kafka if that's available

**Comment by yakkomajuri** [comment]
done @Twixes

File: src/utils/logs-buffer.ts [hunk 5]
```diff
@@ -12,11 +12,19 @@ export class LogsBuffer {
         this.flushTimeout = null
     }
 
-    addLog(log: LogEntryPayload): void {
+    async addLog(log: LogEntryPayload): Promise<void> {
+        // drop into kafka queue immediately as that's aready a buffer
+        if (this.db.kafkaProducer) {
+            await this.db.createPluginLogEntries([log])
+            return
+        }
```
Comment (Twixes) [inline src/utils/logs-buffer.ts:20]: This is kind of confusing. So this way in Kafka/CH mode, we have a `LogsBuffer`, which goes through all the motions as usual… but it's illusory because no data is buffered in it, as the _actual_ buffering is handed of to the Kafka-specific buffer immediately.

File: src/utils/db/db.ts [hunk 6]
```diff
@@ -698,52 +707,73 @@ export class DB {
...
+                    ...[id, team_id, plugin_id, plugin_config_id, timestamp, source, type, message, instance_id],
+                ]
             }
+
+            await this.postgresQuery(
+                `INSERT INTO posthog_pluginlogentry (id, team_id, plugin_id, plugin_config_id, timestamp, source,type, message, instance_id) 
+                VALUES ${valuesString}`,
+                values,
+                'insertPluginLogEntries'
+            )
```
Comment (Twixes) [inline src/utils/db/db.ts:772]: I'd like to ask you to rearchitect this bit. Let's put Kafka and Postgres on equal buffering footing to make this as straightforward as possible – since `queueMessage` is already buffered within a `KafkaProducerWrapper`, we should have something like `queuePostgresInsert` buffered within a `PostgresWrapper`. This way there is each of Kafka and Postgres only ever has a single simple layer of buffering, and that layer is Kafka-/Postgres-specific.

**Review by Twixes — CHANGES_REQUESTED** [review]
Sorry for dragging this on a bit, but my initial point still stands unfortunately 😅

**Comment by yakkomajuri** [comment]
apologies for the lazy updates on the existing system instead of just rearchitecting in the first place :D

@Twixes

File: src/worker/plugins/teardown.ts [hunk 7]
```diff
@@ -14,33 +14,33 @@ export async function teardownPlugins(server: Hub, pluginConfig?: PluginConfig):
...
-                            await server.db.createPluginLogEntry(
+                            await server.db.queuePluginLogEntry({
                                 pluginConfig,
-                                PluginLogEntrySource.System,
-                                PluginLogEntryType.Error,
-                                `Plugin failed to unload (instance ID ${server.instanceId}).`,
-                                server.instanceId
-                            )
+                                source: PluginLogEntrySource.System,
+                                type: PluginLogEntryType.Info,
```
Comment (Twixes) [inline src/worker/plugins/teardown.ts:29]: Why the change to `Info`?

Comment (yakkomajuri) [reply src/worker/plugins/teardown.ts:29]: great catch! probably lingered from a copy paste

**Review by Twixes — APPROVED** [review]
Benissimo
